File: backend/routes/playlists.js
const express = require('express');
const SpotifyService = require('../services/SpotifyService');
const AppleMusicService = require('../services/AppleMusicService');
const YouTubeMusicService = require('../services/YouTubeMusicService');

const router = express.Router();

// Middleware to check authentication
const requireAuth = (platform) => {
  return (req, res, next) => {
    let isAuthenticated = false;
    
    switch (platform) {
      case 'spotify':
        isAuthenticated = !!(req.session.spotifyTokens && req.session.spotifyTokens.expires_at > Date.now());
        break;
      case 'apple-music':
        isAuthenticated = !!(req.session.appleMusicTokens && req.session.appleMusicTokens.expires_at > Date.now());
        break;
      case 'youtube-music':
        isAuthenticated = !!(req.session.youtubeMusicAuth && req.session.youtubeMusicAuth.expires_at > Date.now());
        break;
    }
    
    if (!isAuthenticated) {
      return res.status(401).json({ error: `Not authenticated with ${platform}` });
    }
    
    next();
  };
};

// Get playlists from Spotify
router.get('/spotify', requireAuth('spotify'), async (req, res) => {
  try {
    const spotifyService = new SpotifyService(req.session.spotifyTokens);
    const playlists = await spotifyService.getUserPlaylists();
    res.json(playlists);
  } catch (error) {
    console.error('Error fetching Spotify playlists:', error);
    res.status(500).json({ error: 'Failed to fetch Spotify playlists' });
  }
});

// Get specific playlist tracks from Spotify
router.get('/spotify/:playlistId', requireAuth('spotify'), async (req, res) => {
  try {
    const { playlistId } = req.params;
    const spotifyService = new SpotifyService(req.session.spotifyTokens);
    const tracks = await spotifyService.getPlaylistTracks(playlistId);
    res.json(tracks);
  } catch (error) {
    console.error('Error fetching Spotify playlist tracks:', error);
    res.status(500).json({ error: 'Failed to fetch playlist tracks' });
  }
});

// Get playlists from Apple Music
router.get('/apple-music', requireAuth('apple-music'), async (req, res) => {
  try {
    const appleMusicService = new AppleMusicService(req.session.appleMusicTokens);
    const playlists = await appleMusicService.getUserPlaylists();
    res.json(playlists);
  } catch (error) {
    console.error('Error fetching Apple Music playlists:', error);
    res.status(500).json({ error: 'Failed to fetch Apple Music playlists' });
  }
});

// Get specific playlist tracks from Apple Music
router.get('/apple-music/:playlistId', requireAuth('apple-music'), async (req, res) => {
  try {
    const { playlistId } = req.params;
    const appleMusicService = new AppleMusicService(req.session.appleMusicTokens);
    const tracks = await appleMusicService.getPlaylistTracks(playlistId);
    res.json(tracks);
  } catch (error) {
    console.error('Error fetching Apple Music playlist tracks:', error);
    res.status(500).json({ error: 'Failed to fetch playlist tracks' });
  }
});

// Get playlists from YouTube Music
router.get('/youtube-music', requireAuth('youtube-music'), async (req, res) => {
  try {
    const youtubeMusicService = new YouTubeMusicService(req.session.youtubeMusicHeaders);
    const playlists = await youtubeMusicService.getUserPlaylists();
    res.json(playlists);
  } catch (error) {
    console.error('Error fetching YouTube Music playlists:', error);
    res.status(500).json({ error: 'Failed to fetch YouTube Music playlists' });
  }
});

// Get specific playlist tracks from YouTube Music
router.get('/youtube-music/:playlistId', requireAuth('youtube-music'), async (req, res) => {
  try {
    const { playlistId } = req.params;
    const youtubeMusicService = new YouTubeMusicService(req.session.youtubeMusicHeaders);
    const tracks = await youtubeMusicService.getPlaylistTracks(playlistId);
    res.json(tracks);
  } catch (error) {
    console.error('Error fetching YouTube Music playlist tracks:', error);
    res.status(500).json({ error: 'Failed to fetch playlist tracks' });
  }
});

module.exports = router;

File: backend/routes/playlists.py
from fastapi import APIRouter, Request, Depends, Response, HTTPException
from typing import List
import asyncio

from models.playlist import Playlist
from services.spotify_service import SpotifyService
from services.apple_music_service import AppleMusicService
from services.youtube_music_service import YouTubeMusicService

router = APIRouter()

# --- Dependency functions for authentication ---

def get_spotify_service(request: Request) -> SpotifyService:
    token_info = request.session.get('spotify_token')
    if not token_info:
        raise HTTPException(status_code=401, detail="Not authenticated with Spotify")
    return SpotifyService(auth_token=token_info)

def get_apple_music_service(request: Request) -> AppleMusicService:
    token_info = request.session.get('apple_music_token')
    if not token_info:
        raise HTTPException(status_code=401, detail="Not authenticated with Apple Music")
    return AppleMusicService(user_token=token_info['user_token'])

def get_ytm_service(request: Request) -> YouTubeMusicService:
    headers = request.session.get('youtube_music_headers')
    if not headers:
        raise HTTPException(status_code=401, detail="Not authenticated with YouTube Music")
    return YouTubeMusicService(headers=headers)


# --- Data mapping functions ---

def map_spotify_playlist(p: dict) -> Playlist:
    return Playlist(
        id=p['id'],
        title=p['name'],
        description=p.get('description'),
        track_count=p['tracks']['total'],
        source='spotify'
    )

def map_apple_music_playlist(p: dict) -> Playlist:
    attrs = p.get('attributes', {})
    return Playlist(
        id=p['id'],
        title=attrs.get('name'),
        description=attrs.get('description', {}).get('standard'),
        track_count=attrs.get('trackCount', 0),
        source='apple-music'
    )

def map_ytm_playlist(p: dict) -> Playlist:
    return Playlist(
        id=p['playlistId'],
        title=p['title'],
        description=p.get('description'),
        track_count=p.get('count', 0),
        source='youtube-music'
    )

# --- API Endpoints ---

@router.get("/{platform}", response_model=List[Playlist])
async def get_playlists(platform: str, request: Request):
    if platform == "spotify":
        service = get_spotify_service(request)
        playlists_raw = await service.get_user_playlists()
        return [map_spotify_playlist(p) for p in playlists_raw]
        
    elif platform == "apple-music":
        service = get_apple_music_service(request)
        playlists_raw = await service.get_user_playlists()
        return [map_apple_music_playlist(p) for p in playlists_raw]

    elif platform == "youtube-music":
        service = get_ytm_service(request)
        # ytmusicapi is sync, run in thread to not block event loop
        playlists_raw = await asyncio.to_thread(service.get_user_playlists)
        return [map_ytm_playlist(p) for p in playlists_raw]

    raise HTTPException(status_code=404, detail="Platform not supported")

@router.get("/{platform}/{playlist_id}/tracks")
async def get_playlist_tracks(platform: str, playlist_id: str, request: Request):
    # Implementation will be similar, getting the service
    # and then calling get_playlist_tracks, followed by mapping.
    return {"message": f"Tracks for {platform} playlist {playlist_id}"} 